File: components/elements/clock_element.cpp
#include "clock_element.h"

#include <sys/time.h>

namespace esphome::elements {

void ClockElement::draw(display::Display &display) {
  // Center the clock in the middle of the display.
  const int center_x = display.get_width() / 2;
  const int center_y = display.get_height() / 2;
  const int radius = std::min(center_x, center_y);

  // Draw the markers.
  for (int i = 0; i < 60; i++) {
    if (quarter_markers_.visible && i % 15 == 0) {
      display.line_at_angle(
          center_x, center_y, 6 * i, radius * quarter_markers_.start,
          radius * quarter_markers_.end, quarter_markers_.color);
    } else if (hour_markers_.visible && i % 5 == 0) {
      display.line_at_angle(center_x, center_y, 6 * i,
                            radius * hour_markers_.start,
                            radius * hour_markers_.end, hour_markers_.color);
    } else if (minute_markers_.visible) {
      display.line_at_angle(
          center_x, center_y, 6 * i, radius * minute_markers_.start,
          radius * minute_markers_.end, minute_markers_.color);
    }
  }

  // Compute fractional times.
  const auto time = time_->now();
  float seconds = time.second;
  if (second_hand_.smooth) {
    struct timeval tv;
    if (gettimeofday(&tv, nullptr) == 0) {
      seconds += tv.tv_usec / 1000000.0;
    }
  }
  float minutes = time.minute;
  if (minute_hand_.smooth) {
    minutes += seconds / 60.0;
  }
  float hours = time.hour % 12;
  if (hour_hand_.smooth) {
    hours += minutes / 60.0;
  }

  // Draw the hands.
  if (hour_hand_.visible) {
    display.line_at_angle(center_x, center_y, 360.0 * hours / 12.0 - 90.0,
                          radius * hour_hand_.start, radius * hour_hand_.end,
                          hour_hand_.color);
  }
  if (minute_hand_.visible) {
    display.line_at_angle(center_x, center_y, 360.0 * minutes / 60.0 - 90.0,
                          radius * minute_hand_.start,
                          radius * minute_hand_.end, minute_hand_.color);
  }
  if (second_hand_.visible) {
    display.line_at_angle(center_x, center_y, 360.0 * seconds / 60.0 - 90.0,
                          radius * second_hand_.start,
                          radius * second_hand_.end, second_hand_.color);
  }
}

} // namespace esphome::elements

File: components/elements/clock_element.py
import esphome.codegen as cg
import esphome.config_validation as cv
from esphome.components import time
from esphome.const import CONF_ID

from . import shared
from . import element
from . import element_registry
from . import color

CONF_COLOR = "color"
CONF_END = "end"
CONF_HOUR_HAND = "hour_hand"
CONF_HOUR_MARKERS = "hour_markers"
CONF_MINUTE_HAND = "minute_hand"
CONF_MINUTE_MARKERS = "minute_markers"
CONF_QUARTER_MARKERS = "quarter_markers"
CONF_SECOND_HAND = "second_hand"
CONF_SMOOTH = "smooth"
CONF_START = "start"
CONF_TIME = "time"
CONF_VISIBLE = "visible"

# Clock options

ClockOptions = shared.elements_ns.class_("ClockOptions")


def clock_options_schema(start, end, default_color, visible=True, smooth=None):
    schema = cv.Schema(
        {
            cv.Optional(CONF_START, default=start): cv.float_range(-1, 1),
            cv.Optional(CONF_END, default=end): cv.float_range(-1, 1),
            cv.Optional(CONF_COLOR, default=default_color): color.COLOR_SCHEMA,
            cv.Optional(CONF_VISIBLE, default=visible): cv.boolean,
        }
    )
    if smooth is not None:
        schema = schema.extend(
            {
                cv.Optional(CONF_SMOOTH, default=smooth): cv.boolean,
            }
        )
    return schema


async def clock_options_to_code(config):
    return cg.StructInitializer(
        ClockOptions,
        (CONF_START, config.get(CONF_START)),
        (CONF_END, config.get(CONF_END)),
        (CONF_COLOR, await color.color_to_code(config.get(CONF_COLOR))),
        (CONF_VISIBLE, config.get(CONF_VISIBLE)),
        (CONF_SMOOTH, config.get(CONF_SMOOTH) or False),
    )


# Clock Element

ClockElement = shared.elements_ns.class_("ClockElement", element.Element)

CLOCK_ELEMENT_SCHEMA = element.ELEMENT_SCHEMA.extend(
    {
        cv.GenerateID(CONF_ID): cv.declare_id(ClockElement),
        cv.Required(CONF_TIME): cv.use_id(time.RealTimeClock),
        cv.Optional(CONF_MINUTE_MARKERS, default={}): clock_options_schema(
            0.95, 1.00, "#0000ff", visible=False
        ),
        cv.Optional(CONF_HOUR_MARKERS, default={}): clock_options_schema(
            0.90, 1.00, "#0000ff"
        ),
        cv.Optional(CONF_QUARTER_MARKERS, default={}): clock_options_schema(
            0.75, 1.00, "#0000ff"
        ),
        cv.Optional(CONF_SECOND_HAND, default={}): clock_options_schema(
            0.00, 0.75, "#ff0000", smooth=True
        ),
        cv.Optional(CONF_MINUTE_HAND, default={}): clock_options_schema(
            0.00, 0.95, "#ffffff", smooth=False
        ),
        cv.Optional(CONF_HOUR_HAND, default={}): clock_options_schema(
            0.00, 0.66, "#ffffff", smooth=True
        ),
    }
)


async def clock_element_to_code(config, component, parent):
    var = await element.element_to_code(config, component, parent)
    if conf := config.get(CONF_TIME):
        value = await cg.get_variable(conf)
        cg.add(var.set_time(value))
    for key in [
        CONF_MINUTE_MARKERS,
        CONF_HOUR_MARKERS,
        CONF_QUARTER_MARKERS,
        CONF_SECOND_HAND,
        CONF_MINUTE_HAND,
        CONF_HOUR_HAND,
    ]:
        if conf := config.get(key):
            value = await clock_options_to_code(conf)
            cg.add(getattr(var, "set_" + key)(value))
    return var


element_registry.register_element("clock", CLOCK_ELEMENT_SCHEMA, clock_element_to_code)
